### ISSUE TITLE
Potential performance gain

### ISSUE THREAD

**Issue by spenserblack** [spenserblack/gengo#17]
Right now, extensions, shebangs, etc. are assigned to languages, so that the internals roughly mirror the languages file. However, instead of a structure like this:

```
JSON => ["json"]
JSON with Comments => ["json", "jsonc"]
```

there might be potential for much better performance with a structure like this:

```
"json" => [JSON, JSON with Comments]
```

In other words, make the matcher a key to the language, not the other way around.

Filename patterns will need to be a special case, and continue to be iterated over, but the performance hit shouldn't be too bad since interpreters, extensions, and filenames should be much more commonly used.

**Comment by Byron** [comment]
<img width="1180" alt="Screenshot 2023-09-03 at 15 54 30" src="https://github.com/spenserblack/gengo/assets/63622/5c42b9cf-e714-4b69-b7c6-7acfcf47a25f">

In my profile run, when using `git` as source, only about 23% of the runtime are going into the matcher it seems. 87ms, a considerable fraction, is used to deal with `utf8` conversions. If `BStr` could be used, and with it, `[u8]`, this conversion can possibly be skipped. With `gix` this is now easy enough if one would stop using `Path` and rely on `BStr` (along with `/` as separator).

**Comment by spenserblack** [comment]
This has been effectively addressed by #298
